polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -187,15 +240,29 @@ class Offset : public IntegerValueRange {
   Offset(IntegerValueRange range) : IntegerValueRange(range) {}
 
   static Offset zero() {
-    auto zero = ConstantIntRanges::constant(APInt(32, 0));
+    auto zero = ConstantIntRanges::constant(APInt(sizeof(int64_t), 0));
     return Offset(IntegerValueRange(zero));
   }
 
-  static Offset create(Value val, DataFlowSolver &solver) {
-    if (auto *range =
-            solver.lookupState<dataflow::IntegerValueRangeLattice>(val))
-      return Offset(range->getValue());
-    return Offset(IntegerValueRange());
+  static Offset add(const IntegerValueRange &lhs,
+                    const IntegerValueRange &rhs) {
+    if (lhs.isUninitialized() || rhs.isUninitialized())
+      return Offset(IntegerValueRange());
+    return Offset(Helper::getConst<arith::AddIOp>(lhs, rhs));
+  }
+
+  static Offset sub(const IntegerValueRange &lhs,
+                    const IntegerValueRange &rhs) {
+    if (lhs.isUninitialized() || rhs.isUninitialized())
+      return Offset(IntegerValueRange());
+    return Offset(Helper::getConst<arith::SubIOp>(lhs, rhs));
+  }
+
+  static Offset mul(const IntegerValueRange &lhs,
+                    const IntegerValueRange &rhs) {
+    if (lhs.isUninitialized() || rhs.isUninitialized())
+      return Offset(IntegerValueRange());
+    return Offset(Helper::getConst<arith::MulIOp>(lhs, rhs));
   }
```

Comment:
NIT:
```suggestion
  static Offset add(const IntegerValueRange &lhs,
                    const IntegerValueRange &rhs) {
    return create<arith::AddIOp>(lhs, rhs);
  }

  static Offset sub(const IntegerValueRange &lhs,
                    const IntegerValueRange &rhs) {
    return create<arith::SubIOp>(lhs, rhs);
  }
  
  static Offset mul(const IntegerValueRange &lhs,
                    const IntegerValueRange &rhs) {
    return create<arith::MulIOp>(lhs, rhs);
  }

private:
  template <typename OpTy, typename = std::enable_if_t<...>>
  static Offset create(const IntegerValueRange &lhs,
                    const IntegerValueRange &rhs) {
    if (lhs.isUninitialized() || rhs.isUninitialized())
      return Offset(IntegerValueRange());
    return Offset(Helper::getConst<opTy>(lhs, rhs));
  }
```

Something like this would make sense.